File: test_lut_client.py
"""
TCP client for testing lut_server.

Usage:
    python test_lut_client.py [--host 127.0.0.1] [--port 9000]
                              [--lat 32.08] [--lon 34.76] [--alt-agl 10]
                              [--max-range 15000]
                              [--save output.bin]

Without --save: receives the array over TCP and prints a summary.
With --save:    asks the server to write the file directly (server-side path).
"""

import argparse
import socket
import struct
import sys


def send_request(host: str, port: int,
                 lat: float, lon: float, alt: float, max_range: float,
                 save_path: str | None) -> None:

    mode = 1 if save_path else 0

    # Fixed header: 4 doubles + 1 uint8 = 33 bytes
    header = struct.pack("<4dB", lat, lon, alt, max_range, mode)

    if mode == 1:
        encoded = save_path.encode()
        header += struct.pack("<H", len(encoded)) + encoded

    print(f"Connecting to {host}:{port} ...")
    with socket.create_connection((host, port), timeout=120) as s:
        s.sendall(header)
        print(f"Request sent  lat={lat} lon={lon} alt_agl={alt}m max_range={max_range}m  mode={'save' if mode else 'receive'}")

        if mode == 0:
            _receive_and_print(s, max_range)
        else:
            _receive_status(s, save_path)


def _recvall(s: socket.socket, n: int) -> bytes:
    buf = bytearray()
    while len(buf) < n:
        chunk = s.recv(n - len(buf))
        if not chunk:
            raise ConnectionError(f"Connection closed after {len(buf)}/{n} bytes")
        buf.extend(chunk)
    return bytes(buf)


def _receive_and_print(s: socket.socket, max_range: float) -> None:
    print("Waiting for array dimensions ...")
    dims = _recvall(s, 8)
    az_count, range_count = struct.unpack("<2I", dims)
    total_cells = az_count * range_count
    total_bytes = total_cells * 4

    print(f"Array shape : int32[{az_count}][{range_count}]  ({total_bytes / 1024 / 1024:.1f} MB)")
    print(f"Receiving {total_cells:,} int32 values ...")

    raw = _recvall(s, total_bytes)
    cells = struct.unpack(f"<{total_cells}i", raw)

    # ── Spot checks ──────────────────────────────────────────────────────────
    print("\n── Spot checks ─────────────────────────────────────────────")

    def elev(az_idx: int, range_idx: int) -> int:
        return cells[az_idx * range_count + range_idx]

    checks = [
        (0,   0,   "az=0.0°  range=0m    (radar position)"),
        (12,  100, "az=1.2°  range=1500m"),
        (900, 200, "az=90.0° range=3000m (East)"),
        (az_count - 1, range_count - 1, f"az={((az_count-1)*0.1):.1f}°  range={(range_count-1)*15}m (last cell)"),
    ]
    for ai, ri, label in checks:
        if ai < az_count and ri < range_count:
            print(f"  [{ai:4d}][{ri:4d}]  {label:45s}  elev={elev(ai, ri):6d} m")

    # ── Statistics ───────────────────────────────────────────────────────────
    non_zero = sum(1 for v in cells if v != 0)
    min_elev = min(cells)
    max_elev = max(cells)
    print(f"\n── Statistics ──────────────────────────────────────────────")
    print(f"  Min elev : {min_elev} m")
    print(f"  Max elev : {max_elev} m")
    print(f"  Non-zero : {non_zero:,} / {total_cells:,} cells ({100*non_zero/total_cells:.1f}%)")
    print("\nOK")


def _receive_status(s: socket.socket, save_path: str) -> None:
    status_byte = _recvall(s, 1)
    status = struct.unpack("B", status_byte)[0]
    if status == 0:
        print(f"Server saved file: {save_path}  OK")
    else:
        print(f"Server reported error saving: {save_path}", file=sys.stderr)
        sys.exit(1)


def main() -> None:
    p = argparse.ArgumentParser(description=__doc__,
                                formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--host",      default="127.0.0.1")
    p.add_argument("--port",      type=int,   default=9000)
    p.add_argument("--lat",       type=float, default=31.860342182745416)
    p.add_argument("--lon",       type=float, default=34.91973316662749)
    p.add_argument("--alt-agl",   type=float, default=0.0,
                   help="Radar height above local ground in meters (AGL)")
    p.add_argument("--max-range", type=float, default=15000.0)
    p.add_argument("--save",      default=None,
                   metavar="PATH", help="Ask server to save binary file at this path")
    args = p.parse_args()

    send_request(args.host, args.port,
                 args.lat, args.lon, args.alt_agl, args.max_range,
                 args.save)


if __name__ == "__main__":
    main()

File: server/test_lut_client.cpp
/**
 * TCP client for testing lut_server in C++.
 * Mirrors the logic of test_lut_client.py.
 */

#include <iostream>
#include <string>
#include <vector>
#include <cstring>
#include <cstdint>
#include <iomanip>

#ifdef _WIN32
    #include <winsock2.h>
    #include <ws2tcpip.h>
    #pragma comment(lib, "ws2_32.lib")
#else
    #include <sys/socket.h>
    #include <arpa/inet.h>
    #include <unistd.h>
    #include <netdb.h>
#endif

// Fixed header structure (33 bytes)
#pragma pack(push, 1)
struct LutRequestHeader {
    double  lat;
    double  lon;
    double  alt;
    double  max_range;
    uint8_t mode; // 0 = receive, 1 = server-side save
};
#pragma pack(pop)

void cleanup_sockets() {
#ifdef _WIN32
    WSACleanup();
#endif
}

bool recv_all(int sock, char* buffer, size_t length) {
    size_t total_received = 0;
    while (total_received < length) {
        int received = recv(sock, buffer + total_received, (int)(length - total_received), 0);
        if (received <= 0) return false;
        total_received += received;
    }
    return true;
}

int main(int argc, char* argv[]) {
    std::string host = "127.0.0.1";
    int port = 9000;
    double lat = 31.860342;
    double lon = 34.919733;
    double alt_agl = 0.0;
    double max_range = 15000.0;
    std::string save_path = "";

    // Simple arg parsing
    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];
        if (arg == "--host" && i + 1 < argc) host = argv[++i];
        else if (arg == "--port" && i + 1 < argc) port = std::stoi(argv[++i]);
        else if (arg == "--lat" && i + 1 < argc) lat = std::stod(argv[++i]);
        else if (arg == "--lon" && i + 1 < argc) lon = std::stod(argv[++i]);
        else if (arg == "--alt-agl" && i + 1 < argc) alt_agl = std::stod(argv[++i]);
        else if (arg == "--max-range" && i + 1 < argc) max_range = std::stod(argv[++i]);
        else if (arg == "--save" && i + 1 < argc) save_path = argv[++i];
    }

#ifdef _WIN32
    WSADATA wsaData;
    if (WSAStartup(MAKEWORD(2, 2), &wsaData) != 0) {
        std::cerr << "WSAStartup failed\n";
        return 1;
    }
#endif

    int sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0) {
        std::cerr << "Socket creation failed\n";
        cleanup_sockets();
        return 1;
    }

    sockaddr_in server_addr{};
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    inet_pton(AF_INET, host.c_str(), &server_addr.sin_addr);

    std::cout << "Connecting to " << host << ":" << port << " ...\n";
    if (connect(sock, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        std::cerr << "Connection failed\n";
        cleanup_sockets();
        return 1;
    }

    LutRequestHeader header{ lat, lon, alt_agl, max_range, (uint8_t)(save_path.empty() ? 0 : 1) };
    send(sock, (const char*)&header, sizeof(header), 0);

    if (header.mode == 1) {
        uint16_t path_len = (uint16_t)save_path.length();
        send(sock, (const char*)&path_len, sizeof(path_len), 0);
        send(sock, save_path.c_str(), path_len, 0);
    }

    std::cout << "Request sent. Mode: " << (header.mode == 1 ? "save" : "receive") << "\n";

    if (header.mode == 0) {
        uint32_t az_count, range_count;
        if (!recv_all(sock, (char*)&az_count, 4) || !recv_all(sock, (char*)&range_count, 4)) {
            std::cerr << "Failed to receive dimensions\n";
        } else {
            std::cout << "Array shape: int32[" << az_count << "][" << range_count << "]\n";
            size_t total_cells = (size_t)az_count * range_count;
            std::vector<int32_t> cells(total_cells);
            
            std::cout << "Receiving " << total_cells << " cells (" << (total_cells * 4.0 / 1024 / 1024) << " MB) ...\n";
            
            bool success = true;
            for (size_t i = 0; i < total_cells; ++i) {
                int32_t height_mtr;
                if (!recv_all(sock, (char*)&height_mtr, sizeof(int32_t))) {
                    success = false;
                    break;
                }
                cells[i] = height_mtr;
            }

            if (success) {
                std::cout << "\n── Spot checks ─────────────────────────────────────────────\n";
                auto check = [&](int ai, int ri, const char* label) {
                    if (ai < (int)az_count && ri < (int)range_count) {
                        int32_t val = cells[ai * range_count + ri];
                        std::cout << "  [" << std::setw(4) << ai << "][" << std::setw(4) << ri << "]  " 
                                  << std::left << std::setw(45) << label << " elev=" << val << " m\n";
                    }
                };

                check(0, 0, "az=0.0°  range=0m    (radar position)");
                check(12, 100, "az=1.2°  range=1500m");
                check(900, 200, "az=90.0° range=3000m (East)");
                check(az_count - 1, range_count - 1, "last cell");

                int32_t min_e = cells[0], max_e = cells[0];
                size_t non_zero = 0;
                for (auto v : cells) {
                    if (v < min_e) min_e = v;
                    if (v > max_e) max_e = v;
                    if (v != 0) non_zero++;
                }
                std::cout << "\n── Statistics ──────────────────────────────────────────────\n";
                std::cout << "  Min elev : " << min_e << " m\n";
                std::cout << "  Max elev : " << max_e << " m\n";
                std::cout << "  Non-zero : " << non_zero << " / " << total_cells << " (" << (100.0 * non_zero / total_cells) << "%)\n";
                std::cout << "\nOK\n";
            } else {
                std::cerr << "Failed to receive data\n";
            }
        }
    } else {
        uint8_t status;
        if (recv_all(sock, (char*)&status, 1)) {
            if (status == 0) std::cout << "Server saved file: " << save_path << " OK\n";
            else std::cerr << "Server reported error saving file\n";
        }
    }

#ifdef _WIN32
    closesocket(sock);
#else
    close(sock);
#endif
    cleanup_sockets();
    return 0;
}
